### PR TITLE
Tighten public surface of Expression class.

### LINKED ISSUE #4722: Public Surface of System.Linq.Expressions.Expression exposing too much?
In the code base, the methods [`DebugView`](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L222), [`ReturnReadOnly<T>`](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L245), [`ReturnReadOnly`](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L263), [`ReturnObject<T>`](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L276), [`RequiresCanRead`](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L281), [`RequiresCanRead` (another overload)](https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs#L286), are all public.

They are not documented, nor public in any of the other .Net versions. I can't see any trace of a decision to make them public here, and the blame record suggests they were imported as such. With the possible exception of `DebugView`, they don't seem particularly useful to outside code or if they are they should perhaps be defined elsewhere. Also, `RequiresCanWrite` remains private and should surely be public if `RequiresCanRead` is.

This looks like it was introduced by accident in a refactoring of some of these methods into `ExpressionUtils` that happened in between the release of .NET 4.6 and the code being brought into dotnet/corefx (or in a different branch, etc.).

If I'm correct in deducing that their access is a mistake, removing the access to them would be a breaking change, but by the same token should perhaps be done as soon as possible before more dependencies on them appear.

### PR DESCRIPTION
**Pull request by JonHanna**
Fixes #4722 

Please see that issue wrt whether this should be done or not.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @VSadov

**Comment by JonHanna** [comment]
Being no longer public, the overload of `RequiresCanRead` that takes an `IEnumerable<Expression>` can be optimised to make direct use of the `IReadOnlyList<Expression>` it invariably is called with, instead of testing-and-casting to get to the enumerator-free path.

`RequiresCanWrite` meanwhile could always be optimised to avoid unnecessary set-and-test paths.

Added another commit for this.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs [hunk 1]
```diff
@@ -219,8 +220,10 @@ public override string ToString()
         /// Creates a <see cref="String"/> representation of the Expression.
         /// </summary>
         /// <returns>A <see cref="String"/> representation of the Expression.</returns>
-        public string DebugView
+        internal string DebugView
         {
```
Comment (VSadov) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs]: perhaps should be private. It is private in System.Core

**Comment by VSadov** [comment]
These are likely public because they are used from System.Dynamic. Dynamic used to live in System.Core, but now it lives in a separate assembly.
Does System.Dynamic build with these changes?

It would be a good change to remove these from public surface, even if we need to duplicate some of the helpers on the Dynamic side, but I do not know how viral the change could be. One helper may require copying over another and so on...

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Expression.cs]: Agreed. Done.

**Comment by JonHanna** [comment]
@VSadov well the compilation of the assemblies covered by the CI certainly build. Is there more to dynamic outside of this?

As I said in #4722, it's certainly a breaking change, but if it is in error then perhaps better to break sooner rather than later.

**Comment by JonHanna** [comment]
@dotnet-bot test this, please.

**Comment by VSadov** [comment]
ReturnObject has an entry in the rd.xml  .  I will remove that after trying on .NET Native.

**Pull request merged**